From 856e3ce862c605b2c941658094f3db1e0da6879f Mon Sep 17 00:00:00 2001
From: Thierry Quervel <thierry.quervel@intel.com>
Date: Fri, 28 Sep 2012 10:32:54 +0200
Subject: Restore BT A2DP device availability as soon as possible

BZ: 50998 112163
[CLEAN]

This patch resets the FORCE_NO_BT_A2DP flag to FORCE_NONE
when wired headset or headphone is unplugged (if no WIDI
device available).
This allows to swap directly on BT A2DP without transient
playback on speaker.

Also contains a fix made in I552ade7c59a5e39def2aafb7d2c11f5213ed4e01

Category: aosp improvement
Domain: AudioComms-AudioPolicy
Origin: internal
Upstream-Candidate: no, needs rework
Change-Id: Ia6f5dc09b2e03cc4d1ccd61bd370ed99a0c860e0
Orig-Change-Id: If90c5bcfdd140e477051dcb4c871cac7c822ad78
Signed-off-by: Thierry Quervel <thierry.quervel@intel.com>
Signed-off-by: ptonnetx <pierre-louisx.tonnet@intel.com>
---
 audio/AudioPolicyManagerBase.cpp |   16 ++++++++++++++++
 1 files changed, 16 insertions(+), 0 deletions(-)

diff --git a/audio/AudioPolicyManagerBase.cpp b/audio/AudioPolicyManagerBase.cpp
index ad9b7c5..fc61397 100644
--- a/audio/AudioPolicyManagerBase.cpp
+++ b/audio/AudioPolicyManagerBase.cpp
@@ -136,6 +136,22 @@ status_t AudioPolicyManagerBase::setDeviceConnectionState(audio_devices_t device
             // remove device from available output devices
             mAvailableOutputDevices = (audio_devices_t)(mAvailableOutputDevices & ~device);
 
+            // Change policy in the case of wsHS unplugging with BT A2DP connected:
+            // at that moment, the FORCE_NO_BT_A2DP flag is set; that unfortunately leads to select
+            // temporarily the speaker as output until setForceUse(FORCE_NONE) is called
+            // by AudioService.
+            // Check that BT A2DP connection is available, and that there is no WIDI
+            // device, otherwise keep the flag to prioritize WIDI over BT A2DP
+            if ((device == AudioSystem::DEVICE_OUT_WIRED_HEADSET) ||
+                (device == AudioSystem::DEVICE_OUT_WIRED_HEADPHONE)){
+                if ((mForceUse[AudioSystem::FOR_MEDIA] == AudioSystem::FORCE_NO_BT_A2DP) &&
+                    (getA2dpOutput() != 0) &&
+                    !(mAvailableOutputDevices & AudioSystem::DEVICE_OUT_WIDI)) {
+                    ALOGD("Disabling priority over BT A2DP for MEDIA streams");
+                    mForceUse[AudioSystem::FOR_MEDIA] = AudioSystem::FORCE_NONE;
+                }
+            }
+
             checkOutputsForDevice(device, state, outputs);
             if (mHasA2dp && audio_is_a2dp_device(device)) {
                 // handle A2DP device disconnection
-- 
1.7.4.1

